In [ ]:
import pandas as pd
import numpy as np
import re

## Load 

In [105]:
customers_dirty = pd.read_csv('../data/raw/customers_dirty.csv')
customers_dirty2 = pd.read_csv('../data/raw/customers_dirty2.csv')
customers_dirty3 = pd.read_csv('../data/raw/customers_dirty3.csv')

In [60]:
customers_dirty

customer_id      full_name                      email signup_date country  \
0         3001     Jean Morel     jean.morel@example.com  2025-01-10      FR   
1         3002    Alice Petit     alice.petitexample.com  2025-01-12      FR   
2         3003    Carlos Diaz    carlos.diaz@example.com  not_a_date      ES   
3         3004  Lucie Bernard  lucie.bernard@example.com  2025-02-01      fr   
4         3005    Mohamed Ali    mohamed.ali@example.com  2025-02-10      DE   
5         3006    Emma Dupont    emma.dupont@example.com  2025-02-15      UK   
6         3007      Tom Leroy      tom.leroy@example.com  2025-03-01      US   
7         3008    Sarah Klein    sarah.klein@example.com  2025-03-05      DE   
8         3009     Anna Rossi     anna.rossi@example.com  2025-03-10      IT   
9         3010     Jean Morel     jean.morel@example.com  2025-01-10      FR   

     age  last_purchase_amount loyalty_tier  
0   42.0                 120.5         GOLD  
1   35.0                  80.0       SILVER  
2   29.0                  60.0         GOLD  
3   17.0                  25.0       BRONZE  
4   31.0                 -10.0       SILVER  
5    NaN                  45.0         GOLD  
6  150.0                  90.0         GOLD  
7   28.0                   0.0       SILVER  
8   34.0                  70.0         GOLD  
9   42.0                 120.5         GOLD

Email:
- alice.petitexample.com : email invalides

Signup_date:
- not_a_date: pas format date

Country: 
- fr minuscule: pas standard

age:
- NaN : valeur null
- 150 : arrière*5 papi?

last_purchase_amount:
- -10 : montant négatif


In [61]:
customers_dirty2

customer_id      full_name                      email   signup_date  \
0         2001    Paul Martin    paul.martin@example.com    2025-03-10   
1         2002  Laura Bernard   laura.bernardexample.com    2025-03-12   
2         2003    Marco Rossi    marco.rossi@example.com    2025-02-29   
3         2004  Sophie Durand  sophie.durand@example.com    2025-01-15   
4         2005    Lucas Petit    lucas.petit@example.com    2024-12-01   
5         2006  Anna Kowalski         anna.k@example.com    2025-04-01   
6         2007     Emma Leroy     emma.leroy@example.com    2025-03-05   
7         2008    Noah Dupont    noah.dupont@example.com  invalid_date   
8         2009  Clara Meunier  clara.meunier@example.com    2025-03-01   
9         2010    Paul Martin    paul.martin@example.com    2025-03-10   

  country   age  last_purchase_amount loyalty_tier  
0      FR  45.0                 250.0         GOLD  
1      FR  32.0                  80.0       SILVER  
2      IT  38.0                 120.0         GOLD  
3  France  16.0                  40.0       BRONZE  
4      fr  27.0                  -5.0       SILVER  
5      PL  29.0                  60.0         GOLD  
6      FR   NaN                  30.0       SILVER  
7      FR  34.0                  75.0         GOLD  
8      US  31.0                   0.0     PLATINUM  
9      FR  45.0                 250.0         GOLD

Email:
- laura.bernardexample.com : email invalides

Signup_date:
- invalid_date : pas format date

Country: 
- France, fr: pas standard

age:
- NaN : valeur nulle

last_purchase_amount:
- -5 : montant négatif

In [62]:
customers_dirty3

customer_id      full_name                     email signup_date country  \
0          1001   Marie Dupont  marie.dupont@example.com  2025-11-03      FR   
1          1002     John Smith        john.smith@example  2025-10-15  France   
2          1003                        alice@example.com  2025-13-01      FR   
3          1004   Pedro García  pedro.garcia@example.com  not_a_date      ES   
4          1005         Li Wei        li.wei@example.com  2025-09-12      CN   
5          1006   Chloé Martin  chloe.martin@example.com  2025-09-12      fr   
6          1007  Ahmed Ben Ali  ahmed.benali@example.com  2024-02-29      TN   
7          1008   Sarah Connor  sarah.connor@example.com  2025-08-01      UK   
8          1009    Jean Dupont   jean.dupont@example.com  2025-07-20      FR   
9          1010    Anna Müller  anna.mueller@example.com  2025-07-18      DE   
10         1011   Marie Dupont  marie.dupont@example.com  2025-11-03      FR   
11         1012            Tom                     test@  2025-06-30      US   
12         1013    Inès Dubois   ines.dubois@example.com  2025-06-15     FRA   
13         1014   Carlos Lopez  carlos.lopez@example.com  2025-05-10      ES   
14         1015     Nina Rossi    nina.rossi@example.com  2025-05-10      IT   
15         1016            Zoé           zoe@example.com  2025-04-01      FR   
16         1017            NaN       no_name@example.com  2025-03-22      FR   
17         1018       Jane Doe      jane.doe@example.com  2025-02-30     USA   
18         1019   Émile Moreau  emile.moreau@example.com  2025-01-15      FR   
19         1020    Lucie Petit   lucie.petit@example.com  2025-01-14      FR   

    age  last_purchase_amount loyalty_tier  
0    29                120.50         GOLD  
1    34                 89.90       SILVER  
2    27                 45.00       BRONZE  
3   NaN                  0.00       SILVER  
4    41                -12.00         GOLD  
5    17                 15.00     PLATINUM  
6    25                300.00         GOLD  
7   200                 55.00       SILVER  
8    31                   NaN       BRONZE  
9    28                 19.99       SILVER  
10   29                120.50         GOLD  
11  abc                 10.00       BRONZE  
12   33            9999999.99         GOLD  
13   40                 12.50       SILVER  
14   39                 12.00      UNKNOWN  
15  NaN                  5.00       BRONZE  
16   22                 25.00       SILVER  
17   30                 70.00         GOLD  
18   30                  0.00       BRONZE  
19   30                  0.00       BRONZE

invalid_datEmail:
- john.smith@example, test@ : email invalides

Signup_date:
- not_a_date : pas format date

Country: 
- France, fr, FRA, USA: pas standard
- 200 : impossible
age:
- NaN : valeur nulle
- abc : pas un age

last_purchase_amount:
- -12 : montant négatif
- NaN : valeur manquante

loyalty_tier:
- UNKNOWN : invalid

## analyse and describe

In [63]:
customers_dirty.info()

<class 'pandas.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           10 non-null     int64  
 1   full_name             10 non-null     str    
 2   email                 10 non-null     str    
 3   signup_date           10 non-null     str    
 4   country               10 non-null     str    
 5   age                   9 non-null      float64
 6   last_purchase_amount  10 non-null     float64
 7   loyalty_tier          10 non-null     str    
dtypes: float64(2), int64(1), str(5)
memory usage: 772.0 bytes


In [64]:
customers_dirty.describe()

customer_id         age  last_purchase_amount
count     10.00000    9.000000             10.000000
mean    3005.50000   45.333333             60.100000
std        3.02765   39.981246             45.546557
min     3001.00000   17.000000            -10.000000
25%     3003.25000   29.000000             30.000000
50%     3005.50000   34.000000             65.000000
75%     3007.75000   42.000000             87.500000
max     3010.00000  150.000000            120.500000

In [65]:
customers_dirty2.info()

<class 'pandas.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           10 non-null     int64  
 1   full_name             10 non-null     str    
 2   email                 10 non-null     str    
 3   signup_date           10 non-null     str    
 4   country               10 non-null     str    
 5   age                   9 non-null      float64
 6   last_purchase_amount  10 non-null     float64
 7   loyalty_tier          10 non-null     str    
dtypes: float64(2), int64(1), str(5)
memory usage: 772.0 bytes


In [66]:
customers_dirty2.describe()

customer_id        age  last_purchase_amount
count     10.00000   9.000000             10.000000
mean    2005.50000  33.000000             90.000000
std        3.02765   9.082951             92.225569
min     2001.00000  16.000000             -5.000000
25%     2003.25000  29.000000             32.500000
50%     2005.50000  32.000000             67.500000
75%     2007.75000  38.000000            110.000000
max     2010.00000  45.000000            250.000000

In [67]:
customers_dirty3.info()

<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           20 non-null     int64  
 1   full_name             19 non-null     str    
 2   email                 20 non-null     str    
 3   signup_date           20 non-null     str    
 4   country               20 non-null     str    
 5   age                   18 non-null     str    
 6   last_purchase_amount  19 non-null     float64
 7   loyalty_tier          20 non-null     str    
dtypes: float64(1), int64(1), str(6)
memory usage: 1.4 KB


In [68]:
customers_dirty3.describe()

customer_id  last_purchase_amount
count     20.00000          1.900000e+01
mean    1010.50000          5.263625e+05
std        5.91608          2.294146e+06
min     1001.00000         -1.200000e+01
25%     1005.75000          7.500000e+00
50%     1010.50000          1.999000e+01
75%     1015.25000          7.995000e+01
max     1020.00000          1.000000e+07

---
Fiche qualitée:
- email valid: check avec regex email (si invalid : si uniquement @ manquant l'ajouter, sinon nan)
- date format iso : datetime64, ceux qui sont invalids on met la mediane
- code pays standard : mapping
- 16 =< age <= 100 : int64,  ceux qui sont a nan on met à 16 ou 100 
- montant >= 0
- loyalty_tier : remplacer UNKNOWN par BRONZE